unison-cli/src/Unison/LSP/Types.hs

```diff
@@ -105,14 +112,31 @@ data FileAnalysis = FileAnalysis
   { fileUri :: Uri,
     fileVersion :: FileVersion,
     lexedSource :: LexedSource,
+    tokenMap :: IM.IntervalMap Position Lexer.Lexeme,
     parsedFile :: Maybe (UF.UnisonFile Symbol Ann),
     typecheckedFile :: Maybe (UF.TypecheckedUnisonFile Symbol Ann),
     notes :: Seq (Note Symbol Ann),
     diagnostics :: IntervalMap Position [Diagnostic],
-    codeActions :: IntervalMap Position [CodeAction]
+    codeActions :: IntervalMap Position [CodeAction],
+    fileSummary :: Maybe FileSummary
   }
   deriving (Show)
 
+-- | A file that parses might not always type-check, but often we just want to get as much
+-- information as we have available. This provides a type where we can summarize the
+-- information available in a Unison file.
+--
+-- If the file typechecked then all the Ref Ids and types will be filled in, otherwise
+-- they will be Nothing.
+data FileSummary = FileSummary
```

Comment:
As the doc says, this is a really helpful type for allowing the LSP endpoint implementations to not really need to care whether the current file type-checks or not, they can just use as much information as is available without needing to implement the endpoint twice.